keras_cv/models/densenet.py

```diff
@@ -0,0 +1,306 @@
+    )(x)
+    x = layers.AveragePooling2D(2, strides=2, name=name + "_pool")(x)
+    return x
+
+
+def conv_block(x, growth_rate, name):
+    """A building block for a dense block.
+
+    Args:
+      x: input tensor.
+      growth_rate: float, growth rate at dense layers.
+      name: string, block label.
+
+    Returns:
+      Output tensor for the block.
+    """
+    x1 = layers.BatchNormalization(axis=BN_AXIS, epsilon=1.001e-5, name=name + "_0_bn")(
+        x
+    )
+    x1 = layers.Activation("relu", name=name + "_0_relu")(x1)
+    x1 = layers.Conv2D(4 * growth_rate, 1, use_bias=False, name=name + "_1_conv")(x1)
+    x1 = layers.BatchNormalization(axis=BN_AXIS, epsilon=1.001e-5, name=name + "_1_bn")(
+        x1
+    )
+    x1 = layers.Activation("relu", name=name + "_1_relu")(x1)
+    x1 = layers.Conv2D(
+        growth_rate, 3, padding="same", use_bias=False, name=name + "_2_conv"
+    )(x1)
+    x = layers.Concatenate(axis=BN_AXIS, name=name + "_concat")([x, x1])
+    return x
+
+
+def _get_name(blocks):
+    if blocks == [6, 12, 24, 16]:
+        return "densenet121"
+    if blocks == [6, 12, 32, 32]:
+        return "densenet169"
+    if blocks == [6, 12, 48, 32]:
+        return "densenet201"
+    return "densenet"
+
+
+def _apply_classifier(x, classes, classifier_activation):
+    x = layers.GlobalAveragePooling2D(name="avg_pool")(x)
+    x = layers.Dense(classes, activation=classifier_activation, name="predictions")(x)
+    return x
+
+
+def _apply_pooling_layer(x, pooling):
+    print(x.shape)
+    if pooling == "avg":
+        return layers.GlobalAveragePooling2D(name="avg_pool")(x)
+    if pooling == "max":
+        return layers.GlobalMaxPooling2D(name="max_pool")(x)
+    return x
+
+
+def DenseNet(
+    blocks,
+    include_preprocessing,
+    include_top=True,
+    weights=None,
+    input_tensor=None,
+    input_shape=(None, None, 3),
+    pooling=None,
+    classes=1000,
+    classifier_activation="softmax",
+):
+    """Instantiates the DenseNet architecture.
+
+    Reference:
+    - [Densely Connected Convolutional Networks (CVPR 2017)](
+        https://arxiv.org/abs/1608.06993)
+
+    This function returns a Keras DenseNet model.
+
+    For transfer learning use cases, make sure to read the
+    [guide to transfer learning & fine-tuning](
+      https://keras.io/guides/transfer_learning/).
+
+    Args:
+      blocks: numbers of building blocks for the four dense layers.
+      include_preprocessing: whether or not to Rescale the inputs.
+        If set to True, inputs will be passed through a
+        `Rescaling(1/255.0)` layer.
+      include_top: whether to include the fully-connected
+        layer at the top of the network.
+      weights: one of `None` (random initialization), or a pretrained
+        checkpoint.
+      input_tensor: optional Keras tensor
+        (i.e. output of `layers.Input()`)
+        to use as image input for the model.
+      input_shape: optional shape tuple, defaults to (None, None, 3).
```

Comment:
existing keras applications has some assertion about imagenet weights and the input shape, we might want to put something here as well (eg the weights should work with input shape.)

Reply:
We don't offer any weights at the moment, so we can add this when we add weights back in.